transports/tcp/src/lib.rs

```diff
@@ -365,30 +367,45 @@ where
         tracing::debug!(address=%socket_addr, "dialing address");
 
         let socket = self
+            .config
             .create_socket(socket_addr, opts.port_use)
             .map_err(TransportError::Other)?;
 
+        let mut bound_to_local = false;
+
         match self.port_reuse.local_dial_addr(&socket_addr.ip()) {
             Some(socket_addr) if opts.port_use == PortUse::Reuse => {
                 tracing::trace!(address=%addr, "Binding dial socket to listen socket address");
                 socket
                     .bind(&socket_addr.into())
                     .map_err(TransportError::Other)?;
+                bound_to_local = true;
             }
             _ => {}
         }
 
-        socket
-            .set_nonblocking(true)
-            .map_err(TransportError::Other)?;
+        let local_config = self.config.clone();
 
         Ok(async move {
             // [`Transport::dial`] should do no work unless the returned [`Future`] is polled. Thus
             // do the `connect` call within the [`Future`].
-            match socket.connect(&socket_addr.into()) {
-                Ok(()) => {}
-                Err(err) if err.raw_os_error() == Some(libc::EINPROGRESS) => {}
-                Err(err) if err.kind() == io::ErrorKind::WouldBlock => {}
+            let socket = match socket.connect(&socket_addr.into()) {
+                Ok(()) => socket,
+                Err(err) if err.raw_os_error() == Some(libc::EINPROGRESS) => socket,
+                Err(err) if err.kind() == io::ErrorKind::WouldBlock => socket,
+                Err(err) if err.kind() == io::ErrorKind::AddrNotAvailable && bound_to_local => {
+                    // The socket was bound to a local address that is no longer available.
+                    // Retry without binding.
+                    tracing::trace!("Retrying dial without binding to local address");
```

Comment:
This should be a `debug!` log, I think. Let's also say something like:

```suggestion
                    tracing::trace!(connect_addr = %socket_addr, ?bind_addr, "Failed to connect, re-dialing without binding to existing local socket despite instruction to reuse port");
```

Reply:
That's good!